src/vm/exceptionhandling.cpp

```diff
@@ -203,10 +204,23 @@ static inline void UpdatePerformanceMetrics(CrawlFrame *pcfThisFrame, BOOL bIsRe
     ETW::ExceptionLog::ExceptionThrown(pcfThisFrame, bIsRethrownException, bIsNewException);
 }
 
-void ShutdownEEAndExitProcess()
+#ifdef FEATURE_PAL
+static LONG volatile g_termination_triggered = 0;
+
+void HandleTerminationRequest(int terminationExitCode)
 {
-    ForceEEShutdown(SCA_ExitProcessWhenShutdownComplete);
+    // We set a non-zero exit code to indicate the process didn't terminate cleanly.
+    // This value can be changed by the user by setting Environment.ExitCode in the
+    // ProcessExit event. We only start termination on the first SIGTERM signal
+    // to ensure we don't overwrite an exit code already set in ProcessExit.
```

Comment:
What happens if the SIGTERM causes an int-returning Program.Main to exit gracefully and return a non-zero exit code?

I'm aware of aspnet/AspNetCore#6526 which tracks using Environment.ExitCode in the ProcessExit event to set the process exit code to zero when the ASP.NET host exits gracefully.

I'm more concerned about apps that return custom exit codes from Program.Main. Today this works as long as you avoid `dotnet run` as demonstrated in dotnet/corefx#32749. Now that dotnet/cli#10544 has been merged, `dotnet run` shouldn't even be an issue going forward.

Reply:
> What happens if the SIGTERM causes an int-returning Program.Main to exit gracefully and return a non-zero exit code?

The program observing the exit code will assume the application didn't exit gracefully.

> I'm more concerned about apps that return custom exit codes from Program.Main.

The ProcessExit handler must be aware of those. It can behave different based on the Environment.ExitCode value.